File: component.ts
export class Component {
    private state: string;

    constructor(initialState: string) {
        this.state = initialState;
    }

    render() {
        console.log(`Component state: ${this.state}`);
    }

    setState(newState: string) {
        this.state = newState;
        this.render();
    }
}

File: component.ml
(* component.ml *)
type component = {
    mutable state: string;
    render: unit -> unit;
}

let create_component initial_state =
    let render () = Printf.printf "Component state: %s\n" initial_state in
    let set_state new_state =
        initial_state := new_state;
        render ()
    in
    { state = initial_state; render; set_state }

